File: ts/consent/engine.ts
import { parseScope } from "../scopes/parser";

export type Decision =
  | { result: "GRANT_FAST" }
  | { result: "GRANT_LOGGED" }
  | { result: "REQUIRE_APPROVAL" }
  | { result: "REQUIRE_PROOF" }
  | { result: "DENY"; reason: string };

export function evaluateHandshake(
  requestedScopes: string[],
  trustProfile: Set<string>,
  cachedGrants: Set<string>,
  proofsPresent: boolean
): Decision {

  for (const s of requestedScopes) {
    const scope = parseScope(s);
    if (scope.modifiers["DENIED"]) {
      return { result: "DENY", reason: "Explicit deny modifier" };
    }
  }

  if (requestedScopes.every(s => cachedGrants.has(s))) {
    return { result: "GRANT_FAST" };
  }

  if (requestedScopes.every(s => trustProfile.has(s))) {
    return { result: "GRANT_FAST" };
  }

  if (requestedScopes.some(s => s.includes("PAYMENTS") || s.includes("PII"))) {
    return { result: "REQUIRE_APPROVAL" };
  }

  if (!proofsPresent) {
    return { result: "REQUIRE_PROOF" };
  }

  return { result: "GRANT_LOGGED" };
}

File: py/consent/engine.py
from oracleos.scopes.parser import parse_scope

def evaluate_handshake(scopes, trust_profile, cached_grants, proofs_present):

    for s in scopes:
        parsed = parse_scope(s)
        if parsed["modifiers"].get("DENIED"):
            return ("DENY", "Explicit deny modifier")

    if all(s in cached_grants for s in scopes):
        return ("GRANT_FAST", None)

    if all(s in trust_profile for s in scopes):
        return ("GRANT_FAST", None)

    if any("PAYMENTS" in s or "PII" in s for s in scopes):
        return ("REQUIRE_APPROVAL", None)

    if not proofs_present:
        return ("REQUIRE_PROOF", None)

    return ("GRANT_LOGGED", None)